### PR TITLE
React to https://github.com/aspnet/Razor/pull/683

### PR DESCRIPTION
**Pull request by NTaylorMullen**
/cc @dougbu @rynowak

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 1]
```diff
@@ -289,11 +286,7 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
                        .AppendHtml("||document.write(\"");
 
                 // May have no "src" attribute in the dictionary e.g. if Src and SrcInclude were not bound.
-                if (!attributes.ContainsName(SrcAttributeName))
-                {
-                    // Need this entry to place each fallback source.
-                    attributes.Add(new TagHelperAttribute(SrcAttributeName, value: null));
-                }
+                var addSrc = !attributes.ContainsName(SrcAttributeName);
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Is there a reason not to inline this at the if where it's used?

**Comment by ryanbrandenburg** [comment]
Seems alright to me other than a nitpick, but I'll leave the squirreling to those who know the area better.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs [hunk 2]
```diff
@@ -402,7 +398,7 @@ private void EnsureFileVersionProvider()
             }
         }
 
-        private void BuildLinkTag(TagHelperAttributeList attributes, TagHelperContent builder)
+        private void BuildLinkTag(string newHref, TagHelperAttributeList attributes, TagHelperContent builder)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: Just `href`

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: +1 Just need a `bool` variable above the `for (var i = 0; i < attributes.Count; i++)` and set to `true` if a `src` attribute is seen.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs [hunk 3]
```diff
@@ -410,30 +406,35 @@ private void BuildLinkTag(TagHelperAttributeList attributes, TagHelperContent bu
...
                 {
-                    // "href" values come from bound attributes and globbing. So anything but a non-null string is
-                    // unexpected but could happen if another helper targeting the same element does something odd.
-                    // Pass through existing value in that case.
-                    var attributeStringValue = attributeValue as string;
-                    if (attributeStringValue != null)
-                    {
-                        attributeValue = _fileVersionProvider.AddFileVersionToPath(attributeStringValue);
-                    }
+                    continue;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: Why is it :ok: to reorder the attributes in the `LinkTagHelper` but not in the `ScriptTagHelper`? Prefer choice made in the `ScriptTagHelper`.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 4]
```diff
@@ -337,6 +325,20 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
             }
         }
 
+        private void AppendEncodedVersionedSrc(string name, string srcValue, DefaultTagHelperContent builder)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Just `src`

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs [hunk 5]
```diff
@@ -112,7 +112,7 @@ internal MvcRazorHost(IChunkTreeCache chunkTreeCache, RazorPathNormalizer pathNo
                     TagHelperContentGetContentMethodName = nameof(TagHelperContent.GetContent),
                     TagHelperOutputIsContentModifiedPropertyName = nameof(TagHelperOutput.IsContentModified),
                     TagHelperOutputContentPropertyName = nameof(TagHelperOutput.Content),
-                    TagHelperOutputGetChildContentAsyncMethodName = nameof(TagHelperExecutionContext.GetChildContentAsync)
+                    TagHelperOutputGetChildContentAsyncMethodName = nameof(TagHelperOutput.GetChildContentAsync)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs]: Nice catch

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: Fixed

**Comment by NTaylorMullen** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 6]
```diff
@@ -289,12 +286,6 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
                        .AppendHtml("||document.write(\"");
 
                 // May have no "src" attribute in the dictionary e.g. if Src and SrcInclude were not bound.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Move this comment down to right above `if  (addSrc)` line.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs [hunk 7]
```diff
@@ -306,8 +303,7 @@ private void BuildGlobbedLinkTags(TagHelperAttributeList attributes, TagHelperCo
                     continue;
                 }
 
-                attributes.SetAttribute(HrefAttributeName, url);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: Does the `<link>` tag helper also have a case where the `href` attribute may not be bound? If so, the previous `SetAttribute(HrefAttributeName, url)` would have done the Right Thing:tm: and add the attribute if it wasn't already present. But new code in `BuildLinkTag()` expects `attributes` will _always_ contain `href`.

Bottom line: May need an `addHref` and handling similar to that in the `<script>` tag helper. Unless of course the case is impossible here.

**Comment by dougbu** [comment]
:watch: for https://github.com/aspnet/Mvc/pull/4064#discussion-diff-52395144

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs (hunk 7 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: :+1:

**Comment by NTaylorMullen** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs [hunk 8]
```diff
@@ -402,38 +398,60 @@ private void EnsureFileVersionProvider()
...
 
+        private void AppendEncodedVersionedHref(string hrefName, string hrefValue, TagHelperContent builder)
+        {
+            if (AppendVersion == true)
+            {
+                hrefValue = _fileVersionProvider.AddFileVersionToPath(hrefValue);
+            }
+
+            // attribute.Key ("src") does not need to be JavaScript-encoded.
+            var encodedValue = JavaScriptEncoder.Encode(hrefValue);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: Remove this line. The `<link>` element's `href` attribute value never needs to be JavaScript encoded.

**Comment by dougbu** [comment]
:shipit: but _definitely_ with the fix mentioned [here](https://github.com/aspnet/Mvc/pull/4064#discussion-diff-52408090). `<link>` tag helper broken in this PR without that. (Surprisingly the `<link>` and `<script>` tag helpers do have a few differences :smiling_imp:)

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs (hunk 8 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: :+1:.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs [hunk 9]
```diff
@@ -402,38 +398,57 @@ private void EnsureFileVersionProvider()
...
+            // May have no "href" attribute in the dictionary e.g. if Href and HrefInclude were not bound.
+            if (addHref)
+            {
+                AppendEncodedVersionedHref(HrefAttributeName, href, builder);
             }
 
             builder.AppendHtml("/>");
         }
 
+        private void AppendEncodedVersionedHref(string hrefName, string hrefValue, TagHelperContent builder)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: No longer specially-`Encoded`. Suggest removing that from the name.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs [hunk 10]
```diff
@@ -402,38 +398,57 @@ private void EnsureFileVersionProvider()
...
+                    AppendAttribute(attribute.Name, attribute.Value, builder);
                 }
+            }
 
-                builder
-                    .AppendHtml(attribute.Name)
-                    .AppendHtml("=\"")
-                    .Append(HtmlEncoder, attributeValue)
-                    .AppendHtml("\" ");
+            // May have no "href" attribute in the dictionary e.g. if Href and HrefInclude were not bound.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: What does the state of the dictionary have to do with `asp-href-include` and the `HrefInclude` property? That attribute contributes possible values for the `href`, not something in the dictionary. Unless I'm missing something, this should just be `if Href was not bound`.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 11]
```diff
@@ -337,6 +325,25 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
...
+            string srcValue,
+            TagHelperContent builder,
+            bool javaScriptSrc = true)
+        {
+            if (AppendVersion == true)
+            {
+                srcValue = _fileVersionProvider.AddFileVersionToPath(srcValue);
+            }
+
+            if (javaScriptSrc)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: This should not be conditional. Appending the version could add characters that must be encoded.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 12]
```diff
@@ -316,20 +307,17 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
...
-                            // attribute.Key ("src") does not need to be JavaScript-encoded.
-                            var encodedValue = JavaScriptEncoder.Encode(attributeValue);
-
-                            AppendAttribute(builder, attribute.Name, encodedValue, escapeQuotes: true);
+                            addSrc = false;
+                            AppendEncodedVersionedSrc(attribute.Name, src, builder);
                         }
                     }
 
+                    // May have no "src" attribute in the dictionary e.g. if Src and SrcInclude were not bound.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Again, end with `if Src was not bound.`

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 13]
```diff
@@ -337,6 +325,25 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
             }
         }
 
+        private void AppendEncodedVersionedSrc(
+            string srcName,
+            string srcValue,
+            TagHelperContent builder,
+            bool javaScriptSrc = true)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Give this a better name e.g. `generateForDocumentWrite` and remove default value.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs (hunk 10 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs]: Legacyy comment. Removed.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs (hunk 11 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Talked off line.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Never mind

**Comment by dougbu** [comment]
:shipit: (still) though naming is hard.

**Pull request merged**